contrib/zmq/zmq_sub.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+// Example client code making use of Bitcoin's ZMQ pubsub functionality. Only specialized handling
+// for blockhash is implemented, which retrieves the raw block over RPC and then prints it.
+//
+// You can test this by starting bitcoind on regtest
+//
+//   ./src/bitcoind -zmqpubhashblock=tcp://127.0.0.1:38332 -rpcuser=foo -rpcpassword=bar -regtest -debug=1
+//
+// and running this script with
+//
+//   go get github.com/btcsuite/btcd
+//   go get github.com/pebbe/zmq4
+//   BTC_USER=foo BTC_PASS=bar BTC_RPCPORT=18443 ZMQ_ENDPOINT=tcp://127.0.0.1:38332 go run zmq_sub.go
+//
+// then generating some blocks with
+//
+//   ./src/bitcoin-cli -regtest -rpcuser=foo -rpcpassword=bar generate 1
+//
+
+package main
+
+import (
+	"encoding/hex"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/rpcclient"
+	"github.com/btcsuite/btcd/wire"
+	zmq "github.com/pebbe/zmq4"
+)
+
+var rpc *rpcclient.Client
+
+// The ZMQ address to subscribe to.
+var zmqEndpointStr = getEnv("ZMQ_ENDPOINT", "tcp://127.0.0.1:28332")
+
+// The Bitcoin host address (for use by RPC)
+var btcHost = getEnv("BTC_HOST", "127.0.0.1")
+
+// The Bitcoin RPC port
+var btcRPCPort = getEnv("BTC_RPCPORT", "8332")
+
+// The Bitcoin RPC username
+var btcRPCUser = os.Getenv("BTC_USER")
+
+// The Bitcoin RPC password
+var btcRPCPass = os.Getenv("BTC_PASS")
+
+// Return the value of an environment variable if it's defined, otherwise return fallback.
+func getEnv(key, fallback string) string {
+	if value, found := os.LookupEnv(key); found {
+		return value
+	}
+	return fallback
+}
+
+// Listen for Bitcoin notifications over ZMQ. Push matching messages into the provided channels
+// upon receipt.
+func listenForNotification(
+	notifyBlockHash chan<- string,
+	notifyBlockRaw chan<- string,
+	notifyTxHash chan<- string,
+	notifyTxRaw chan<- string,
+	quit <-chan bool,
+) {
+	subscriber, _ := zmq.NewSocket(zmq.SUB)
+	defer subscriber.Close()
+	subscriber.Connect(zmqEndpointStr)
+
+	log.Printf("Listening for zmq notifications on %s\n", zmqEndpointStr)
+
+	// Optionally filter by message type
+	//
+	// subscriber.SetSubscribe("hashblock")
+	//
+	subscriber.SetSubscribe("")
+
+	newmsg := make(chan [][]byte)
+
+	go func() {
+		for {
+			msg, err := subscriber.RecvMessageBytes(0)
+			if err != nil {
+				log.Fatal(err)
+				break
+			}
+			newmsg <- msg
+		}
+	}()
+
+	for {
+		select {
+		case msg := <-newmsg:
+			msgType := string(msg[0])
+			encoded := hex.EncodeToString(msg[1])
```

Comment:
Seems strange that you would need to encode payloads as hex strings here, instead of just passing the data directly. Is the concern 0 bytes? https://golang.org/pkg/builtin/#string seems to indicate that arbitrary byte values are fine.